### PR TITLE
feat(helm): add `template` command

### LINKED ISSUE #2755: proposal: Integrate 'template' plugin functionality
The [template plugin](https://github.com/technosophos/helm-template) fills a need for chart developers and testing systems that need to work without requiring an active Tiller or k8s cluster. 

## Problem
The template plugin does not provide parity with Helm with regards to newer features that alter final  output via mechanisms like Import Values and Tags and Conditions. Maintaining the template plugin separately is more work and will often result in an outdated feature set.

## Proposed Solution
We should integrate the template plugin functionality into core so that chart developers have a built-in command to render charts locally and see output that matches what is expected from a `helm install`.

### Implementation
The plugin can be integrated with its functionality more or less "as is" except it could be downstream in the pipeline and render what would normally be the final chart being sent to Tiller. Any new features and flags should "just work" with the template output. 

### Considerations
Much of the Tiller rendering logic will be compiled into the Helm client increasing the size of the binary and increasing maintenance costs due to adding a new (small) feature.

### Command
The command `template` would be added and would render all templates in the final chart (including its sub-charts) to standard out:
```
helm template foo
```
The above command would render all templates in the foo chart and its sub-charts
```
---
# Source: foo/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: RELEASE-NAME-foo
...
```

## Opens
Since the output only contains rendered templates, should we include a `--debug` flag so that developers can see computed values etc?

### PR DESCRIPTION
**Pull request by jascott1**
This adds the functionality from the helm-template plugin to allow the
rendering of templates without Tiller.

Closes #2755

### CODE REVIEW

**Comment by jascott1** [comment]
### Implementation Notes 
The `--verbose` flag was dropped in favor of the `--debug` flag and the "user values, computed values" output is now the same format as that in `helm install --debug`. 

The `--release` flag was renamed to `--name` for parity with `helm install --name`

The template output is still the same as `helm-template` plugin. 

The following `install` flags were not added to `template` since they do not apply to client-side template rendering:

- dry-run
- no-hooks
- replace
- verify
- keyring
- version
- timeout
- wait
- repo
- cert-file
- key-file
- ca-file
- devel

File: cmd/helm/template_test.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+			desc:        "verify --name-template result exists",
+			args:        []string{chartPath, "--name-template", "foobar-{{ b64enc \"abc\" }}-baz"},
+			expectKey:   "subchart1/templates/service.yaml",
+			expectValue: "release-name: \"foobar-YWJj-baz\"",
+		},
+	}
+
+	var buf bytes.Buffer
+	for _, tt := range tests {
+		// capture stdout
```
Comment (nebril) [inline cmd/helm/template_test.go]: It would be nice to use subtests here: https://blog.golang.org/subtests

File: cmd/helm/template.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+		}
+		printRelease(os.Stdout, rel)
+	}
+
+	sortedKeys := make([]string, 0, len(out))
+	for key := range out {
+		sortedKeys = append(sortedKeys, key)
+	}
+	sort.Strings(sortedKeys)
+	// If renderFiles is set, we ONLY print those.
```
Comment (nebril) [inline cmd/helm/template.go]: Should the command return an error when user provides non-existent file with `-x` flag? If yes, I think it doesn't happen now.

**Review by nebril — COMMENTED** [review]
LGTM overall, two small nitpicks in comments.

File: cmd/helm/template.go (hunk 2 above)
Comment (jascott1) [inline cmd/helm/template.go]: Good catch! We shall flog the original author posthaste! : )

**Comment by technosophos** [comment]
As the author of the original plugin, this is looks great to me. I agree with the refactors done here. It's probably safe to merge this early in the 2.7 cycle.

**Comment by jascott1** [comment]
@technosophos One small UX concern with the `-x` for executing specified templates. Currently it is working as in the original pluign:

`helm template ~/mychart -x mychart/templates/service.yaml`. 

Should we consider modifying it to make it relative to chart? 

`helm template ~/mychart -x templates/service.yaml` 

Also it does not currently work with absolute paths for specified templates. Does anyone think that is important? I can see merit in restricting it to the current chart but can also see allowing it to be absolute and exist anywhere (even outside chart).

File: cmd/helm/template.go (hunk 2 above)
Comment (jascott1) [inline cmd/helm/template.go]: Verification of template existence is now implemented with the caveats described below.

**Comment by technosophos** [comment]
Yeah, I like relative to the chart

**Comment by technosophos** [comment]
How soon can we get this rebased and merged?

**Comment by jascott1** [comment]
@technosophos @nebril This needs to be re-reviewed as I had to make changes and perhaps there is a better way. To re-use kindSorter I had to export tiller.Manifest{} and I choose to create an exported wrapper tiller.SortByKind that just uses InstallOrder for sorting so I would not have to keep exporting things.

**Review by technosophos — APPROVED** [review]
Re-reviewed, and it looks good.

**Comment by bacongobbler** [comment]
Are we holding off for the moment or shall we go ahead and merge this? 😊

**Review by bacongobbler — APPROVED** [review]
looks good to me.

**Pull request merged**